src/widgets/modals/ByProgramModal.js

```diff
@@ -0,0 +1,360 @@
+  supplierError: 'No suppliers available',
+  orderTypeError: 'No order types available',
+  periodError: 'No periods available',
+};
+
+const newState = {
+  RESET_ALL: {
+    program: null,
+    supplier: null,
+    period: null,
+    periods: null,
+    orderType: null,
+    orderTypes: null,
+    name: null,
+    currentStepKey: null,
+    isProgramBased: true,
+  },
+  SELECT_PROGRAM: {
+    period: {},
+    periods: null,
+    supplier: {},
+    orderType: {},
+    orderTypes: null,
+    modalIsOpen: false,
+  },
+  SELECT_ORDER_TYPE: { period: {}, periods: null, modalIsOpen: false },
+  SELECT_NAME: { modalIsOpen: false },
+  SELECT_PERIOD: { modalIsOpen: false },
+  SELECT_SUPPLIER: { modalIsOpen: false },
+};
+
+export class ByProgramModal extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      // Values used for deriving current UI state
+      steps: null,
+      modalIsOpen: false,
+      isProgramBased: true,
+      currentStepKey: null,
+      // New requisition or stocktake values
+      program: null,
+      supplier: null,
+      orderType: null,
+      period: null,
+      name: null,
+      // Data stored for performance
+      storeTag: null,
+      programs: null,
+      suppliers: null,
+      orderTypes: null,
+      periods: null,
+    };
+  }
+
+  // Store unchanging select modal options and set the default
+  // steps array.
+  componentDidMount = () => {
+    const { settings, database } = this.props;
+    const programs = getAllPrograms(settings, database);
+    const suppliers = database.objects('Name').filtered('isSupplier = $0', true);
+    const steps = this.setCurrentSteps();
+    this.setState({ programs, suppliers, currentStepKey: steps[0].key });
+  };
+
+  onConfirmRequisition = () => {
+    const { onConfirm } = this.props;
+    const { supplier: otherStoreName, orderType } = this.state;
+    onConfirm({ ...this.state, otherStoreName, orderType: orderType && orderType.name });
+    this.setState({ ...newState.RESET_ALL }, () => this.setCurrentSteps());
+  };
+
+  onToggleChange = () => {
+    const { isProgramBased } = this.state;
+    this.setState({ ...newState.RESET_ALL, isProgramBased: !isProgramBased }, () =>
+      this.setCurrentSteps(true)
+    );
+  };
+
+  // Open a modal - in case it is an earlier step in the process,
+  // set the current step to the step number corresponding with
+  // the onPress event.
+  onModalTransition = ({ key: currentStepKey }) => {
+    if (currentStepKey) this.setState({ modalIsOpen: true, currentStepKey });
+    else this.setState({ modalIsOpen: false });
+  };
+
+  getSequentialStepsProps = () => {
+    const { name } = this.state;
+
+    const getBaseProps = key => ({
+      name: this.state[key] && this.state[key].name,
+      placeholder: localization[`${key}Title`],
+      key,
+      error: this.state[`${key}s`] ? this.state[`${key}s`].length === 0 : false,
+      errorText: localization[`${key}Error`],
```

Comment:
`${key}s`, `${key}Title`, `${key}Error` is pretty crude, you may think so yourself.
Letting this be just a comment, not requirement, this maybe the only example of how sequential steps component is used, we should think about refactor when we have more time, and some headspace.

Reply:
had to 5ee672eca0a586924679acdb5fd39e4def8c5942